src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,348 @@
+                    let rest_child = RestChild::Right;
+                    return (current_node_index, rest_child);
+                }
+            }
+            BvhNode::Leaf { .. } => {
+                self.has_node = false;
+                (current_node_index, RestChild::None)
+            }
+        }
+    }
+
+    /// Attempt to move to the farthest to ray node child of the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the node [`Aabb`], `has_node` will become false.
+    fn move_furthest(&mut self) -> (usize, RestChild) {
+        let current_node_index = self.node_index;
+        match self.bvh.nodes[current_node_index] {
+            BvhNode::Node {
+                child_l_index,
+                ref child_l_aabb,
+                child_r_index,
+                ref child_r_aabb,
+                ..
+            } => {
+                let (_, left_dist) = self.ray.intersection_slice_for_aabb(child_l_aabb);
+                let (_, right_dist) = self.ray.intersection_slice_for_aabb(child_r_aabb);
+
+                if left_dist < T::zero() {
+                    if right_dist < T::zero() {
+                        // no intersections with any child
+                        self.has_node = false;
+                        (current_node_index, RestChild::None)
+                    } else {
+                        // have intersection only with right child
+                        self.has_node = true;
+                        self.node_index = child_r_index;
+                        let rest_child = RestChild::None;
+                        (current_node_index, rest_child)
+                    }
+                } else if right_dist < T::zero() {
+                    // have intersection only with left child
+                    self.has_node = true;
+                    self.node_index = child_l_index;
+                    let rest_child = RestChild::None;
+                    (current_node_index, rest_child)
+                } else if left_dist < right_dist {
+                    // right is furthest than left
+                    self.has_node = true;
+                    self.node_index = child_r_index;
+                    let rest_child = RestChild::Left;
+                    return (current_node_index, rest_child);
+                } else {
+                    // left is furthest than right
+                    self.has_node = true;
+                    self.node_index = child_l_index;
+                    let rest_child = RestChild::Right;
+                    return (current_node_index, rest_child);
+                }
+            }
+            BvhNode::Leaf { .. } => {
+                self.has_node = false;
+                (current_node_index, RestChild::None)
+            }
+        }
+    }
+
+    /// Attempt to move to the rest not visited child of the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the node [`Aabb`], `has_node` will become false.
```

Comment:
I'm not really sure what the first sentence is supposed to mean. Could you try to rephrase it?
```suggestion
    /// Attempt to move to the rest not visited child of the current node.
    /// If it is a leaf, or the [`Ray`] does not intersect the node [`Aabb`], `has_node` will become `false`.
```